util/grpc/interceptor.go

```diff
@@ -50,6 +58,46 @@ var (
 	}
 )
 
+// SetVersionHeaderUnaryServerInterceptor returns a new unary server interceptor that sets the argo-version header
+func SetVersionHeaderUnaryServerInterceptor(version wfv1.Version) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		m, err := handler(ctx, req)
+		if err == nil {
+			// Don't set header if there was an error because attackers could use it to find vulnerable Argo servers
+			err := grpc.SetHeader(ctx, metadata.Pairs(ArgoVersionHeader, version.Version))
+			if err != nil {
+				log.Warnf("Failed to set header '%s': %s", ArgoVersionHeader, err)
+			}
+		}
+		return m, err
```

Comment:
will this return the original error, or the second error from `SetHeader`? we should return the original one, as the warning is sufficient for the header

in most languages, I would think it's the second right now, but you are using `:=` instead of `=` in the second one which would also normally have a lint error in most languages for shadowed variables with the same name.

Reply:
It'll return the original header, because Go is [lexically scoped](https://go.dev/ref/spec#Declarations_and_scope), and the `if err == nil {` block establishes a new scope. I pushed eb08e7665c092a11cbb24f49953d6fdf02d3b1d8 to add tests cases for this: https://github.com/argoproj/argo-workflows/blob/b958ca759f930eefedb46c584fdaad752a0b4dc6/util/grpc/interceptor_test.go#L126-L135

Reply:
Right that's what I thought; normally linters disallow that as variable "shadowing". For clarity and better readability, I would suggest rewriting to `origErr` and `err` like in https://github.com/argoproj/argo-workflows/pull/11674

Reply:
Since the equivalent code is used in the function below this, this could be split into an encapsulated helper.

Reply:
Okay, I pushed a8a283fd3ecbab8c00fa3f421e41419966212d74 to rename `err` to `origErr`.

Also, this can't be refactored into a helper because `grpc.SetHeader()` doesn't work for stream servers, because there's no context available to the inerceptor.

Reply:
Ah I thought there was a diff I noticed earlier that I missed this time around, thanks for pointing that out again!